lib/pos_epoch.go

```diff
@@ -0,0 +1,182 @@
+package lib
+
+import (
+	"bytes"
+	"github.com/dgraph-io/badger/v3"
+	"github.com/pkg/errors"
+)
+
+//
+// TYPE
+//
+
+type EpochEntry struct {
+	EpochNumber      uint64
+	FinalBlockHeight uint64
+}
+
+func (epochEntry *EpochEntry) Copy() *EpochEntry {
+	return &EpochEntry{
+		EpochNumber:      epochEntry.EpochNumber,
+		FinalBlockHeight: epochEntry.FinalBlockHeight,
+	}
+}
+
+func (epochEntry *EpochEntry) RawEncodeWithoutMetadata(blockHeight uint64, skipMetadata ...bool) []byte {
+	var data []byte
+	data = append(data, UintToBuf(epochEntry.EpochNumber)...)
+	data = append(data, UintToBuf(epochEntry.FinalBlockHeight)...)
+	return data
+}
+
+func (epochEntry *EpochEntry) RawDecodeWithoutMetadata(blockHeight uint64, rr *bytes.Reader) error {
+	var err error
+
+	// EpochNumber
+	epochEntry.EpochNumber, err = ReadUvarint(rr)
+	if err != nil {
+		return errors.Wrapf(err, "EpochEntry.Decode: Problem reading EpochNumber: ")
+	}
+
+	// FinalBlockHeight
+	epochEntry.FinalBlockHeight, err = ReadUvarint(rr)
+	if err != nil {
+		return errors.Wrapf(err, "EpochEntry.Decode: Problem reading FinalBlockHeight: ")
+	}
+
+	return err
+}
+
+func (epochEntry *EpochEntry) GetVersionByte(blockHeight uint64) byte {
+	return 0
+}
+
+func (epochEntry *EpochEntry) GetEncoderType() EncoderType {
+	return EncoderTypeEpochEntry
+}
+
+//
+// UTXO VIEW UTILS
+//
+
+func (bav *UtxoView) GetCurrentEpochEntry() (*EpochEntry, error) {
+	var epochEntry *EpochEntry
+	var err error
+
+	// First, check the UtxoView.
+	epochEntry = bav.CurrentEpochEntry
+	if epochEntry != nil {
+		return epochEntry.Copy(), nil
+	}
+
+	// If not found, check the database.
+	epochEntry, err = DBGetCurrentEpochEntry(bav.Handle, bav.Snapshot)
+	if err != nil {
+		return nil, errors.Wrapf(err, "UtxoView.GetCurrentEpoch: problem retrieving EpochEntry from db: ")
+	}
+	if epochEntry != nil {
+		// Cache in the UtxoView.
+		bav.CurrentEpochEntry = epochEntry.Copy()
+	}
+	return epochEntry, nil
+}
+
+func (bav *UtxoView) GetCurrentEpochNumber() (uint64, error) {
+	epochEntry, err := bav.GetCurrentEpochEntry()
+	if err != nil {
+		return 0, errors.Wrapf(err, "UtxoView.GetCurrentEpochNumber: ")
+	}
+	if epochEntry == nil {
+		return 0, errors.New("UtxoView.GetCurrentEpochNumber: no CurrentEpochEntry found")
+	}
+	return epochEntry.EpochNumber, nil
+}
+
+func (bav *UtxoView) SetCurrentEpochEntry(epochEntry *EpochEntry, blockHeight uint64) error {
+	// This function should only ever be called from the OnEpochEnd hook.
+	if epochEntry == nil {
+		return errors.New("UtxoView.SetCurrentEpochEntry: called with nil EpochEntry")
+	}
+
+	// TODO: is this what we should do here?
+	if bav.Snapshot != nil {
```

Comment:
not sure if this is the correct place for this logic. I'll need to think about its proper place a little more. I think it should be handled in `FlushToDbWithTxn`